sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -359,6 +368,27 @@ as we can guarantee we only need at most two buffers for any call to write (two
         }
     }
 
+    private Mono<Response<BlockBlobItem>> determineUploadFullOrChunked(final Flux<ByteBuffer> data,
+        final Function<Flux<ByteBuffer>, Mono<Response<BlockBlobItem>>> path1,
+        final BiFunction<Flux<ByteBuffer>, Long, Mono<Response<BlockBlobItem>>> path2) {
+        final int[] predicateCount = { 4 * Constants.MB };
+        final boolean[] usePath1 = { true };
+
+        return data
+            .filter(ByteBuffer::hasRemaining)
+            .doOnEach(signal -> usePath1[0] = !signal.isOnComplete())
+            .bufferUntil(buffer -> {
+                predicateCount[0] -= buffer.remaining();
+                return predicateCount[0] < 0;
+            }, true)
+            .next()
+            .flatMap(bufferedData -> usePath1[0]
+                ? path1.apply(data)
+                : path2.apply(Flux.fromIterable(bufferedData), bufferedData.stream()
+                .map(buffer -> (long) buffer.remaining())
+                .reduce(0L, Long::sum)));
```

Comment:
This is cool @alzimmermsft. :) .

I small question 

Do we still need this code?

```java
bufferedData.stream()
     .map(buffer -> (long) buffer.remaining())
     .reduce(0L, Long::sum)
```

if we do `4 * Constants.MB + predicateCount[0]` then that should give us total size of bytes in `bufferedData` right?

Reply:
Correct, could just use `4 * Constants.MB - predicateCount[0]` to get the difference between the 4MB check and what data has been streamed through the Flux.

Reply:
actually we're skipping the last ByteBuffer from buffering but predicateCount[0] includes the last one as well so `4 * Constants.MB + predicateCount[0]` won't work always.